### PR TITLE
Data Access Pattern in 0.5

### PR DESCRIPTION
**Pull request by Evizero**
Implementation or continuation of #13 #14 #15 

The code and tests are now in place and ready for review for anyone interested.

I will leave this open at least until I wrote a basic readthedocs documentation, at which point I will merge unless some things come up.

### CODE REVIEW

**Comment by Evizero** [comment]
@oxinabox I updated `oversample` and `undersample` to reflect my plan for labeled data. See this commit: https://github.com/JuliaML/MLDataUtils.jl/pull/16/commits/1461624da1c70cd4e9b875e3d9722719916c0d08

This introduces a new function `target([f], data)` which returns the, well, target(s) of some given data object.

- By default `target(data)` is the identity function. So if you just pass it a single array, it assumes that this is the target array.

- The optional `f` is a function that specifies how the targets are extracted from `data`. By default it is the identity function. This parameter is useful for custom types that include the target somehow (such as DataFrames, see docstring of commit).

- if `data` is a `Tuple` then the convention is the the last element of the tuple contains the target, and thus the function `target` is applied to `data[end]`. This arbitrary convention is imposed by this package. We already treat tuples special as a "group-observations-together-type", so may as well go a step further.

My changes are not set in stone. Discussion very welcome

File: src/accesspattern/dataiterator.jl [hunk 1]
```diff
@@ -254,17 +254,17 @@ function RandomBatches{T,O}(data::T, size::Int, obsdim::O)
...
+RandomBatches{T}(data::T, size::Int; obsdim = default_obsdim(data)) =
     RandomBatches(data, size, obs_dim(obsdim))
 
-RandomBatches(data, size::Int, count::Int; obsdim=default_obsdim(data)) =
+RandomBatches{T}(data::T, size::Int, count::Int; obsdim = default_obsdim(data)) =
     RandomBatches(data, size, count, obs_dim(obsdim))
 
 # convenience constructor.
-RandomBatches(data, size::Int, ::Void, obsdim) =
+RandomBatches{T}(data::T, size::Int, ::Void, obsdim) =
```
Comment (oxinabox) [inline src/accesspattern/dataiterator.jl]: Why have you added `{T}` everywhere?

Comment (Evizero) [reply src/accesspattern/dataiterator.jl]: For the strangest reason. On my mackbook on 0.5 dispatch did not work as I would have assumed. Without it `RandomBatches(some_array, 10)` did not trigger https://github.com/JuliaML/MLDataUtils.jl/blob/329400a80685ef91a7971708aebbbd8ffe7b0453/src/accesspattern/dataiterator.jl#L138 but instead invoked https://github.com/JuliaML/MLDataUtils.jl/blob/329400a80685ef91a7971708aebbbd8ffe7b0453/src/accesspattern/dataiterator.jl#L133. I don't yet understand why

Comment (oxinabox) [reply src/accesspattern/dataiterator.jl]: Might be a good SO question, for the future reference of others.
(and me.)

File: docs/index.rst [hunk 2]
```diff
@@ -12,11 +12,13 @@ context.
 Where to begin?
 ----------------
 
-If this is the first time you consider using MLDataUtils for your
-machine learning related experiments, make sure to check out the
-"Getting Started" section; specifically "How to ...?", which
-lists some of most common scenarios and how to approach them
-using the functionality provided by this package.
+/bin/bash: :w: command not found
```
Comment (oxinabox) [inline docs/index.rst]: This is not what was intended
I am sure.

Comment (Evizero) [reply docs/index.rst]: what the

**Comment by oxinabox** [comment]
Woot!

**Comment by Evizero** [comment]
Don't get too excited! The code is functional but the documentation still unfinished. I am preparing to split the package apart as discussed in #29

**Comment by oxinabox** [comment]
It is merged.
The docs are pretty, if not 100% complete, and it has a very intuitive design.
Please tag a v0.1.0 at least here.
So I can list it as a dependency.
Even if you break it down in #29

**Comment by Evizero** [comment]
yes ok, fair enough. Doesn't hurt to have it functional here first before splitting apart.

I'll tag it before the end of this week. Might as well finish rudimentary documentation for all new aspects here.

**Comment by Evizero** [comment]
I just tagged. May take a bit to get into METADATA, as the diff is a bit much to look through.

In general you may be better off to directly depend on MLDataPattern, unless you require DataFrames

**Pull request merged**